aptos-move/aptos-vm/src/transaction_arg_validation.rs

```diff
@@ -83,22 +101,133 @@ pub(crate) fn validate_combine_signer_and_txn_args<S: MoveResolverExt>(
+        if !idxs.contains(&idx) {
+            continue;
+        }
+        let mut cursor = Cursor::new(&arg[..]);
+        validate_arg(session, ty, &mut cursor, &mut current_size_for_strings)?;
+    }
+    Ok(())
+}
+
+// Validate a single arg. A Cursor is used to walk the serialized arg manually and correctly.
+// Only Strings and nested vector of them are valiadated.
+fn validate_arg<S: MoveResolverExt>(
+    session: &SessionExt<S>,
+    ty: &Type,
+    cursor: &mut Cursor<&[u8]>,
+    current_size_for_strings: &mut usize,
+) -> Result<(), VMStatus> {
+    use move_deps::move_vm_types::loaded_data::runtime_types::Type::*;
+    Ok(match ty {
+        Vector(inner) => {
+            // get the vector length and iterate over each element
+            let mut len = get_len(cursor)?;
+            *current_size_for_strings += len;
+            if *current_size_for_strings > MAX_STRING_ARGS_SIZE {
+                return Err(VMStatus::Error(StatusCode::FAILED_TO_DESERIALIZE_ARGUMENT));
+            }
+            while len > 0 {
+                validate_arg(session, inner, cursor, current_size_for_strings)?;
+                len -= 1;
+            }
+        }
+        // only strings are validated, and given we are here only if one was present
+        // (`is_valid_txn_arg`), this match arm must be for a string
+        Struct(idx) | StructInstantiation(idx, _) => {
+            // load the struct name, we use `expect()` because that check was already
+            // performed in `is_valid_txn_arg`
+            let st = session
+                .get_struct_type(*idx)
+                .expect("unreachable, type must exist");
+            let full_name = format!("{}::{}", st.module.short_str_lossless(), st.name);
+            // load the serialized string
+            let len = get_len(cursor)?;
+            *current_size_for_strings += len;
+            if *current_size_for_strings > MAX_STRING_ARGS_SIZE {
+                return Err(VMStatus::Error(StatusCode::FAILED_TO_DESERIALIZE_ARGUMENT));
+            }
+            let mut s = vec![0u8; len];
+            cursor
+                .read_exact(&mut s)
+                .map_err(|_| VMStatus::Error(StatusCode::FAILED_TO_DESERIALIZE_ARGUMENT))?;
+            // validate the struct value, we use `expect()` because that check was already
+            // performed in `is_valid_txn_arg`
+            let option_validator = ALLOWED_STRUCTS
+                .get(&full_name)
+                .expect("unreachable: struct must be allowed");
+            if let Some(validator) = option_validator {
+                validator(&s)?;
+            }
+        }
+        // this is unreachable given the check in `is_valid_txn_arg` and the
+        // fact we collect all arguments that involve strings and we validate
+        // them and them only
+        Bool | U8 | U64 | U128 | Address | Signer | Reference(_) | MutableReference(_)
```

Comment:
I see, so we're still not evaluating the cursor... so we could have actually ended earlier if we evaluate the cursor length here too. The primary issue was not the size of the actual size of the txn, it was that we didn't properly mock a deserialization.

Reply:
We should at least take into consideration size_of for the curent_size_for_strings -- most of these are many bytes big.